internal/utils/kruize/kruize_api.go

```diff
@@ -178,15 +178,96 @@ func Update_recommendations(experiment_name string, interval_end_time time.Time)
 
 }
 
-func Is_valid_recommendation(d []kruizePayload.ListRecommendations) bool {
-	if len(d) > 0 {
-		notifications := d[0].Kubernetes_objects[0].Containers[0].Recommendations.Notifications
-		// 112101 is notification code for "Duration Based Recommendations Available".
-		if _, ok := notifications["112101"]; ok {
-			return true
-		} else {
+func Is_valid_recommendation(recommendation kruizePayload.Recommendation, experiment_name string, maxEndTime time.Time) bool {
+
+	validRecommendationCode := "111000"
+	_, recommendationIsValid := recommendation.Notifications[validRecommendationCode]
+	if recommendationIsValid {
+		// Convert the time object to the expected format
+		formattedMaxEndTime := maxEndTime.UTC().Format("2006-01-02T15:04:05.000Z")
+		recommendationData, timeStampisValid := recommendation.Data[formattedMaxEndTime]
+		if !timeStampisValid {
+			log.Error("recommendation not found for endtime: ", formattedMaxEndTime)
+			invalidRecommendation.Inc()
 			return false
 		}
+		LogKruizeErrors(recommendationData, formattedMaxEndTime, experiment_name)
+		return true
+	} else {
+		return false
+	}
+}
+
+func LogKruizeErrors(recommendationData kruizePayload.RecommendationData, formattedMaxEndTime string, experiment_name string) {
+
+	// https://github.com/kruize/autotune/blob/master/design/NotificationCodes.md#detailed-codes
+	errorNotificationCodes := map[string]string{
+		"221001": "ERROR",
+		"221002": "ERROR",
+		"221003": "ERROR",
+		"221004": "ERROR",
+		"223001": "ERROR",
+		"223002": "ERROR",
+		"223003": "ERROR",
+		"223004": "ERROR",
+		"224001": "ERROR",
+		"224002": "ERROR",
+		"224003": "ERROR",
+		"224004": "ERROR",
+	}
```

Comment:
I know mostly we won't keep this! It seems we can have separate files/packages declaring these kinds of error constants.

Reply:
Yes, but this part of the code felt temporary :smile:  so I just left it there.